serverless-cljs-plugin/lumo/build.cljs

```diff
@@ -0,0 +1,195 @@
+(ns lumo.build
+  (:require os fs path archiver
+            [posix-getopt :as go]
+            [clojure.string :as str]
+            [goog.string :as gstr]
+            [goog.string.format]
+            [goog.object :as gobj]
+            [cljs.reader :as reader]
+            [lumo.io :as io]
+            [lumo.build.api]))
+
+(defn guess-project-name
+  "Try to guess the project name from the compiler options.
+  Should be used as last resort."
+  [compiler-opts]
+  (.-name (.parse path (:output-to compiler-opts))))
+
+(comment
+  (def output-dir "target/logpoc/logpoc.js")
+  x)
+
+(defn zip!
+  "Create the zip in output-dir. Return a promise containing the path.
+
+  The dirs will be included in the zip, the compression level defaults
+  to 9."
+  [output-path zip-opts compiler-opts]
+
+  (let [{:keys [compression-level dirs files]
+         :or {compression-level 9}} zip-opts
+        archiver (archiver "zip" #js {:zlib {:level compression-level}})
+        output-stream (.createWriteStream fs output-path)]
+    (js/Promise.
+     (fn [resolve-fn reject-fn]
+       (.on output-stream "close"
+            (fn []
+              (.info js/console "Zip created in %s is %s bytes" output-path (.pointer archiver))
+              (resolve-fn output-path)))
+
+       (.on archiver "warning"
+            (fn [err]
+              (if (= (.-code err) "ENOENT")
+                (.warn js/console (.-message err))
+                (reject-fn err))))
+
+       (.on archiver "error"
+            (fn [err]
+              (.error js/console (.-message err))
+              (reject-fn err)))
+       ;; start the piping of the following
+       (.pipe archiver output-stream)
+       ;; adding dirs
+       (doseq [d dirs]
+         (.directory archiver d))
+       ;; adding files
+       (doseq [f files]
+         (.file archiver f))
+       ;; finalize
+       (.finalize archiver)))))
+
+(defn compile!
+  "Compile ClojureScript using the lumo api."
+  [inputs compiler-opts]
+  (do
+    (.info js/console (gstr/format "Invoking the Lumo compiler w/ inputs %s" (.stringify js/JSON (clj->js inputs))))
+    (lumo.build.api/build
+     (apply lumo.build.api/inputs inputs compiler-opts)
+     compiler-opts)))
+
+(defn generate-index!
+  "Generate the necessary index.js and copy it to target-path, ready for zipping."
+  [target-path functions compiler-options]
+  ;; TODO
+  )
+
+(defn- cljs-lambda-conf!
+  "Return the cljs-lamdba configuration."
+  []
+  (-> "cljs-lambda.edn"
+      io/slurp
+      reader/read-string))
+
+(defn read-conf!
+  "Read and return the configuration map."
+  []
+  ;; TODO read from serverless.yml
+  ;; TODO read from project.clj as safety net?
+  (cljs-lambda-conf!))
+
+(defn compiler-defaults
+  "Generate the compiler options map part for the output."
+  [target-path project-name]
```

Comment:
Well `:optimizations` might be good to have configurable, for dead code elimination. I would not want to restrict that.

About the output, I am still torn, I like the way normal `serveless` does it, prepending even the `stage` in front of the name of the `.js` file. So I think at some point we'll need to configure it...

Also, there is another issue, if we have `output-dir` we force to split the files...otherwise one `.js` file will be generated. No sure here, but can the split impact performance?